File: regions/system/library/go/auth/middleware.go
package auth

import (
	"context"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
)

// contextKey はコンテキストに Claims を格納するためのキー型。
type contextKey string

const (
	// ClaimsContextKey はコンテキストに格納する Claims のキー。
	ClaimsContextKey contextKey = "auth_claims"
)

// AuthMiddleware は gin 用の JWT 認証ミドルウェアを返す。
// Authorization ヘッダーから Bearer トークンを取得し、JWKS 検証を行う。
// 検証成功時は Claims をコンテキストに格納する。
func AuthMiddleware(verifier *JWKSVerifier) gin.HandlerFunc {
	return func(c *gin.Context) {
		tokenString, err := extractBearerToken(c.Request)
		if err != nil {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"error":   "SYS_AUTH_UNAUTHENTICATED",
				"message": "認証が必要です",
			})
			return
		}

		claims, err := verifier.VerifyToken(c.Request.Context(), tokenString)
		if err != nil {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"error":   "SYS_AUTH_INVALID_TOKEN",
				"message": "トークンが無効です",
			})
			return
		}

		// Claims をコンテキストに格納
		ctx := context.WithValue(c.Request.Context(), ClaimsContextKey, claims)
		c.Request = c.Request.WithContext(ctx)
		c.Set(string(ClaimsContextKey), claims)

		c.Next()
	}
}

// RequireRole は指定ロールを必須とする gin ミドルウェアを返す。
// AuthMiddleware の後に使用すること。
func RequireRole(role string) gin.HandlerFunc {
	return func(c *gin.Context) {
		claims, ok := GetClaims(c)
		if !ok {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"error":   "SYS_AUTH_UNAUTHENTICATED",
				"message": "認証が必要です",
			})
			return
		}

		if !HasRole(claims, role) {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
				"error":   "SYS_AUTH_FORBIDDEN",
				"message": "この操作を実行する権限がありません",
			})
			return
		}

		c.Next()
	}
}

// RequirePermission は指定リソース・アクションの権限を必須とする gin ミドルウェアを返す。
// AuthMiddleware の後に使用すること。
func RequirePermission(resource, action string) gin.HandlerFunc {
	return func(c *gin.Context) {
		claims, ok := GetClaims(c)
		if !ok {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"error":   "SYS_AUTH_UNAUTHENTICATED",
				"message": "認証が必要です",
			})
			return
		}

		if !HasPermission(claims, resource, action) {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
				"error":   "SYS_AUTH_FORBIDDEN",
				"message": "この操作を実行する権限がありません",
			})
			return
		}

		c.Next()
	}
}

// RequireTierAccess は指定 Tier へのアクセスを必須とする gin ミドルウェアを返す。
// AuthMiddleware の後に使用すること。
func RequireTierAccess(tier string) gin.HandlerFunc {
	return func(c *gin.Context) {
		claims, ok := GetClaims(c)
		if !ok {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"error":   "SYS_AUTH_UNAUTHENTICATED",
				"message": "認証が必要です",
			})
			return
		}

		if !HasTierAccess(claims, tier) {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
				"error":   "SYS_AUTH_TIER_FORBIDDEN",
				"message": "このTierへのアクセス権がありません",
			})
			return
		}

		c.Next()
	}
}

// GetClaims は gin.Context から Claims を取得する。
func GetClaims(c *gin.Context) (*Claims, bool) {
	val, exists := c.Get(string(ClaimsContextKey))
	if !exists {
		return nil, false
	}
	claims, ok := val.(*Claims)
	return claims, ok
}

// GetClaimsFromContext は context.Context から Claims を取得する。
func GetClaimsFromContext(ctx context.Context) (*Claims, bool) {
	val := ctx.Value(ClaimsContextKey)
	if val == nil {
		return nil, false
	}
	claims, ok := val.(*Claims)
	return claims, ok
}

// extractBearerToken は HTTP リクエストから Bearer トークンを取得する。
func extractBearerToken(r *http.Request) (string, error) {
	authHeader := r.Header.Get("Authorization")
	if authHeader == "" {
		return "", ErrMissingToken
	}

	parts := strings.SplitN(authHeader, " ", 2)
	if len(parts) != 2 || !strings.EqualFold(parts[0], "Bearer") {
		return "", ErrInvalidAuthHeader
	}

	token := strings.TrimSpace(parts[1])
	if token == "" {
		return "", ErrMissingToken
	}

	return token, nil
}

// エラー定義。
var (
	ErrMissingToken      = &AuthError{Code: "SYS_AUTH_MISSING_TOKEN", Message: "Authorization ヘッダーがありません"}
	ErrInvalidAuthHeader = &AuthError{Code: "SYS_AUTH_INVALID_HEADER", Message: "Authorization ヘッダーの形式が不正です"}
)

// AuthError は認証エラーを表す。
type AuthError struct {
	Code    string
	Message string
}

func (e *AuthError) Error() string {
	return e.Message
}

File: regions/system/library/rust/auth/src/middleware.rs
//! axum 用の認証ミドルウェア。

use crate::claims::Claims;
use crate::rbac;
use crate::verifier::{AuthError, JwksVerifier};
use axum::{
    body::Body,
    extract::{Request, State},
    http::StatusCode,
    middleware::Next,
    response::{IntoResponse, Response},
    Json,
};
use serde_json::json;
use std::sync::Arc;

/// ミドルウェアファクトリの戻り値型。
type AuthMiddlewareFuture = std::pin::Pin<
    Box<dyn std::future::Future<Output = Result<Response, AuthErrorResponse>> + Send>,
>;

/// AuthState はミドルウェアが使用する共有状態。
#[derive(Clone)]
pub struct AuthState {
    pub verifier: Arc<JwksVerifier>,
}

/// auth_middleware は JWT 認証ミドルウェア。
/// Authorization ヘッダーから Bearer トークンを取得し、JWKS 検証を行う。
/// 検証成功時は Claims をリクエストエクステンションに格納する。
pub async fn auth_middleware(
    State(state): State<AuthState>,
    mut req: Request<Body>,
    next: Next,
) -> Result<Response, AuthErrorResponse> {
    let token = extract_bearer_token(&req)?;

    let claims = state
        .verifier
        .verify_token(&token)
        .await
        .map_err(AuthErrorResponse::from_auth_error)?;

    req.extensions_mut().insert(claims);

    Ok(next.run(req).await)
}

/// require_role は指定ロールを必須とするミドルウェアファクトリ。
/// auth_middleware の後に使用すること。
pub fn require_role(
    role: &'static str,
) -> impl Fn(Request<Body>, Next) -> AuthMiddlewareFuture + Clone {
    move |req: Request<Body>, next: Next| {
        Box::pin(async move {
            let claims = req
                .extensions()
                .get::<Claims>()
                .ok_or_else(AuthErrorResponse::unauthenticated)?;

            if !rbac::has_role(claims, role) {
                return Err(AuthErrorResponse::forbidden(
                    "この操作を実行する権限がありません",
                ));
            }

            Ok(next.run(req).await)
        })
    }
}

/// require_permission は指定リソース・アクションの権限を必須とするミドルウェアファクトリ。
/// auth_middleware の後に使用すること。
pub fn require_permission(
    resource: &'static str,
    action: &'static str,
) -> impl Fn(Request<Body>, Next) -> AuthMiddlewareFuture + Clone {
    move |req: Request<Body>, next: Next| {
        Box::pin(async move {
            let claims = req
                .extensions()
                .get::<Claims>()
                .ok_or_else(AuthErrorResponse::unauthenticated)?;

            if !rbac::has_permission(claims, resource, action) {
                return Err(AuthErrorResponse::forbidden(
                    "この操作を実行する権限がありません",
                ));
            }

            Ok(next.run(req).await)
        })
    }
}

/// require_tier_access は指定 Tier へのアクセスを必須とするミドルウェアファクトリ。
/// auth_middleware の後に使用すること。
pub fn require_tier_access(
    tier: &'static str,
) -> impl Fn(Request<Body>, Next) -> AuthMiddlewareFuture + Clone {
    move |req: Request<Body>, next: Next| {
        Box::pin(async move {
            let claims = req
                .extensions()
                .get::<Claims>()
                .ok_or_else(AuthErrorResponse::unauthenticated)?;

            if !rbac::has_tier_access(claims, tier) {
                return Err(AuthErrorResponse::tier_forbidden());
            }

            Ok(next.run(req).await)
        })
    }
}

/// リクエストエクステンションから Claims を取得する。
pub fn get_claims(req: &Request<Body>) -> Option<&Claims> {
    req.extensions().get::<Claims>()
}

/// Bearer トークンを Authorization ヘッダーから取得する。
fn extract_bearer_token(req: &Request<Body>) -> Result<String, AuthErrorResponse> {
    let auth_header = req
        .headers()
        .get(http::header::AUTHORIZATION)
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| AuthErrorResponse::from_auth_error(AuthError::MissingToken))?;

    let parts: Vec<&str> = auth_header.splitn(2, ' ').collect();
    if parts.len() != 2 || !parts[0].eq_ignore_ascii_case("Bearer") {
        return Err(AuthErrorResponse::from_auth_error(
            AuthError::InvalidAuthHeader,
        ));
    }

    let token = parts[1].trim();
    if token.is_empty() {
        return Err(AuthErrorResponse::from_auth_error(AuthError::MissingToken));
    }

    Ok(token.to_string())
}

/// AuthErrorResponse は認証エラーの HTTP レスポンス。
#[derive(Debug)]
pub struct AuthErrorResponse {
    pub status: StatusCode,
    pub code: String,
    pub message: String,
}

impl AuthErrorResponse {
    fn from_auth_error(err: AuthError) -> Self {
        match err {
            AuthError::MissingToken | AuthError::InvalidAuthHeader => Self {
                status: StatusCode::UNAUTHORIZED,
                code: "SYS_AUTH_UNAUTHENTICATED".into(),
                message: "認証が必要です".into(),
            },
            AuthError::TokenExpired => Self {
                status: StatusCode::UNAUTHORIZED,
                code: "SYS_AUTH_TOKEN_EXPIRED".into(),
                message: "トークンの有効期限が切れています".into(),
            },
            AuthError::InvalidToken(_) => Self {
                status: StatusCode::UNAUTHORIZED,
                code: "SYS_AUTH_INVALID_TOKEN".into(),
                message: "トークンが無効です".into(),
            },
            AuthError::JwksFetchFailed(_) => Self {
                status: StatusCode::INTERNAL_SERVER_ERROR,
                code: "SYS_AUTH_JWKS_ERROR".into(),
                message: "認証サービスへの接続に失敗しました".into(),
            },
            AuthError::PermissionDenied => Self::forbidden("この操作を実行する権限がありません"),
            AuthError::TierAccessDenied => Self::tier_forbidden(),
        }
    }

    fn unauthenticated() -> Self {
        Self {
            status: StatusCode::UNAUTHORIZED,
            code: "SYS_AUTH_UNAUTHENTICATED".into(),
            message: "認証が必要です".into(),
        }
    }

    fn forbidden(message: &str) -> Self {
        Self {
            status: StatusCode::FORBIDDEN,
            code: "SYS_AUTH_FORBIDDEN".into(),
            message: message.into(),
        }
    }

    fn tier_forbidden() -> Self {
        Self {
            status: StatusCode::FORBIDDEN,
            code: "SYS_AUTH_TIER_FORBIDDEN".into(),
            message: "このTierへのアクセス権がありません".into(),
        }
    }
}

impl IntoResponse for AuthErrorResponse {
    fn into_response(self) -> Response {
        let body = json!({
            "error": self.code,
            "message": self.message,
        });

        (self.status, Json(body)).into_response()
    }
}
